llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumption"
+
+using namespace llvm;
+namespace {
+
+/// get AddRec scalar evolution, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution *SE) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Op));
+}
+
+/// get the SCEV of the loop IV
+std::optional<std::pair<const SCEVAddRecExpr *, int>>
+getLoopCmpIVAddRecExpr(ICmpInst &LoopCmpInstr, ScalarEvolution *SE) {
+
+  const SCEVAddRecExpr *S0 = getAddRecSCEV(LoopCmpInstr.getOperand(0), SE);
+  const SCEVAddRecExpr *S1 = getAddRecSCEV(LoopCmpInstr.getOperand(1), SE);
+
+  if (S0 && S1) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountAssumptions-Warning: Encountered two ADDRec "
+                  "expressions in the Loop Compare instruction. Currently it "
+                  " is expected that the compare consists of the IV and a "
+                  "static bound, that is loop invariant.");
+    return std::nullopt;
+  }
+
+  if (S0)
+    return std::optional(std::make_pair(S0, 0));
+
+  if (S1)
+    return std::optional(std::make_pair(S1, 1));
+
+  return std::nullopt;
+}
+
+ICmpInst *getLoopCmpInst(Loop &CurrentLoop) {
+  if (!CurrentLoop.getLoopLatch())
+    return nullptr;
+
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  // check if the branch instruction is the correct one
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+
+  if (LoopCmpInstr->getPredicate() == CmpInst::ICMP_EQ ||
+      LoopCmpInstr->getPredicate() == CmpInst::ICMP_NE) {
+    LLVM_DEBUG(dbgs() << "LoopIterCountAssumptions-Warning: EQ and NE "
+                         "conditions are currently not supported!");
+    return nullptr;
+  }
+  return LoopCmpInstr;
+}
+
+} // namespace
+
+PreservedAnalyses LoopIterCountAssumptions::run(Loop &CurrentLoop,
+                                                LoopAnalysisManager &AM,
+                                                LoopStandardAnalysisResults &AR,
+                                                LPMUpdater &U) {
```

Comment:
Could you modularize this function a bit? I think it got a bit big :)

Maybe something as simple as 
```
run() {
  ICmpInst *LoopCmpInstr = getLoopCmpInst(CurrentLoop);
  if (!LoopCmpInstr)
    return PreservedAnalyses::all();

  const SCEV *MinIterSCEV = getSafeMinIterSCEV(...)
  if (!MinIterSCEV)
    return PreservedAnalyses::all();

  insertMinIterAssumption(...);
  }
```

Reply:
I stripped down the run function to only the necessary function calls.